jscomp/runtime/caml_obj.ml

```diff
@@ -198,6 +210,19 @@ let rec caml_compare (a : Obj.t) (b : Obj.t) : int =
         else
           let len_a = Bs_obj.length a in
           let len_b = Bs_obj.length b in
+          if len_a = 0 && len_b = 0 && O.is_object a && O.is_object b then
+            begin
+              let keys_a = O.keys a in
+              let keys_b = O.keys b in
+              O.sort(keys_a);
+              O.sort(keys_b);
+              let len_a = O.length keys_a in
+              let len_b = O.length keys_b in
+              let min_len = min len_a len_b in
+              let default_res = len_a - len_b in
+              aux_obj_compare a keys_a b keys_b 0 min_len default_res
+            end
+        else
           if len_a = len_b then
             aux_same_length a b 0 len_a
```

Comment:
shall we put this branch earlier, this branch seems to be hit more often than the previous one? Is it okay to assume `Array.isArray` API exists? (cc @chenglou )
```
if len_a = len_b && is_array a 
```
In that case, we don't need `o.constructor == Object`